runtime/src/serde_snapshot.rs

```diff
@@ -321,19 +351,64 @@ where
     let mut accounts_db = AccountsDb::new_with_config(
         account_paths.to_vec(),
         cluster_type,
-        account_indexes,
+        account_secondary_indexes,
         caching_enabled,
         shrink_ratio,
     );
-    let AccountsDbFields(storage, version, slot, bank_hash_info) = accounts_db_fields;
+    let AccountsDbFields(
+        full_snapshot_storage,
+        full_snapshot_version,
+        full_snapshot_slot,
+        full_snapshot_bank_hash_info,
+    ) = snapshot_accounts_db_fields.full_snapshot_accounts_db_fields;
+
+    let (snapshot_version, snapshot_slot, snapshot_bank_hash_info) = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(
+            |AccountsDbFields(
+                _,
+                incremental_snapshot_version,
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )| {
+                (
+                    *incremental_snapshot_version,
+                    *incremental_snapshot_slot,
+                    incremental_snapshot_bank_hash_info.clone(),
+                )
+            },
+        )
+        .unwrap_or((
+            full_snapshot_version,
+            full_snapshot_slot,
+            full_snapshot_bank_hash_info,
+        ));
+
+    // filter out incremental snapshot storages with slot <= full snapshot slot
+    let incremental_snapshot_storage = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .map_or_else(HashMap::new, |fields| {
+            fields
+                .0
+                .into_iter()
+                .filter(|(slot, _)| *slot > full_snapshot_slot)
+                .collect::<HashMap<_, _>>()
+        });
+
+    assert!(incremental_snapshot_storage.iter().all(|storage_entry| !full_snapshot_storage.contains_key(storage_entry.0)),
+    "There must not be any overlap in the slots of storages between the full snapshot and the incremental snapshot");
```

Comment:
Should this be a hard assert or just another error result?

Reply:
For instance could a malicious validator make a full snapshot for `X` with slots `Y > X` such that those extra slots `Y` collide with slots in future incremental snapshots?

To prevent this, should we also 
1) verify that a full snapshot for `X` only contains slots `<=X`, 
2) and then if we know incremental snapshots based on `X` only contain slots `> X` from this earlier check here: https://github.com/solana-labs/solana/pull/18504/files#r668284384

Then this non-overlapping naturally falls out of the two guarantees

Reply:
Ooo good points. I like this idea and will do.

Reply:
I've changed the assert to return an error instead. Is that sufficient for now?

I originally put in the assert from your comment here: https://github.com/solana-labs/solana/pull/17875#discussion_r652475517

Reply:
Man, what a flip flopper - can't trust this guy with anything 😛